src/box/tuple_compare.cc

```diff
@@ -1447,17 +1469,48 @@ func_index_compare_with_key(struct tuple *tuple, hint_t tuple_hint,
 			    const char *key, uint32_t part_count,
 			    hint_t key_hint, struct key_def *key_def)
 {
-	(void)tuple; (void)key_hint;
+	(void)key_hint;
 	assert(key_def->for_func_index);
 	assert(is_nullable == key_def->is_nullable);
 	const char *tuple_key = tuple_data((struct tuple *)tuple_hint);
 	assert(mp_typeof(*tuple_key) == MP_ARRAY);
 
 	uint32_t tuple_key_count = mp_decode_array(&tuple_key);
-	part_count = MIN(part_count, tuple_key_count);
+	uint32_t cmp_part_count = MIN(part_count, tuple_key_count);
+	cmp_part_count = MIN(cmp_part_count, key_def->part_count);
+	int rc = key_compare_and_skip_parts<is_nullable>(&tuple_key, &key,
+							 cmp_part_count,
+							 key_def);
+	if (rc != 0)
+		return rc;
+	/* Equals if nothing to compare. */
+	if (part_count == cmp_part_count ||
+	    key_def->part_count == cmp_part_count)
+		return 0;
+	/*
+	 * Now we know that tuple_key count is less than key part_count
+	 * and key_def part_count, so let's keep comparing, but with
+	 * original tuple fields. We will compare parts of primary key,
+	 * it cannot contain nullable parts so the code is simplified
+	 * correspondingly. Also, all the key parts, corresponding to
+	 * func key, were already skipped.
+	 */
+	const char *tuple_raw = tuple_data(tuple);
+	struct tuple_format *format = tuple_format(tuple);
+	const uint32_t *field_map = tuple_field_map(tuple);
+	const char *field;
 	part_count = MIN(part_count, key_def->part_count);
-	return key_compare_parts<is_nullable>(tuple_key, key, part_count,
-					      key_def);
+	for (uint32_t i = cmp_part_count; i < part_count; i++) {
+		struct key_part *part = &key_def->parts[i];
+		field = tuple_field_raw_by_part(format, tuple_raw, field_map,
```

Comment:
Please also prepare a patch for [`key_def_raw.c`](https://github.com/tarantool/tarantool-ee/blob/a74aeaf7c6f1526a3aa723ed443f620eeb2cc265/src/box/key_def_raw.c#L242-L247) from the EE repository: currently, it assumes that the function index implementation of `tuple_compare_with_key` doesn't access tuple format, but it isn't true anymore so we need to provide an override.

Reply:
The function may actually be used by a read view in Lua so a stub isn't an option. An override that uses functional key parts only and has an assertion that an extended key definition (with primary key parts) is never used should be fine. I opened a ticket and assigned it to you: https://github.com/tarantool/tarantool-ee/issues/248.

Reply:
[Done](https://github.com/tarantool/tarantool-ee/pull/250).